pkg/controllers/deprovisioning/controller.go

```diff
@@ -0,0 +1,513 @@
+	lastConsolidationState int64
+}
+
+// pollingPeriod that we inspect cluster to look for opportunities to deprovision
+const pollingPeriod = 10 * time.Second
+
+var errCandidateNodeDeleting = fmt.Errorf("candidate node is deleting")
+
+// waitRetryOptions are the retry options used when waiting on a node to become ready or to be deleted
+// readiness can take some time as the node needs to come up, have any daemonset extended resoruce plugins register, etc.
+// deletion can take some time in the case of restrictive PDBs that throttle the rate at which the node is drained
+var waitRetryOptions = []retry.Option{
+	retry.Delay(2 * time.Second),
+	retry.LastErrorOnly(true),
+	retry.Attempts(60),
+	retry.MaxDelay(10 * time.Second), // 22 + (60-5)*10 =~ 9.5 minutes in total
+}
+
+func NewController(clk clock.Clock, kubeClient client.Client, provisioner *provisioning.Provisioner,
+	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster) *Controller {
+	return &Controller{
+		clock:         clk,
+		kubeClient:    kubeClient,
+		cluster:       cluster,
+		provisioner:   provisioner,
+		recorder:      recorder,
+		cloudProvider: cp,
+		emptiness:     &Emptiness{kubeClient: kubeClient, clock: clk, cluster: cluster},
+		consolidation: NewConsolidation(clk, kubeClient, provisioner, cp, recorder, cluster),
+	}
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) controller.Builder {
+	return controller.NewSingletonManagedBy(m).
+		Named("deprovisioning")
+}
+
+func (c *Controller) LivenessProbe(_ *http.Request) error {
+	return nil
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// the last cluster consolidation wasn't able to improve things and nothing has changed regarding
+	// the cluster that makes us think we would be successful now
+	if c.lastConsolidationState == c.cluster.ClusterConsolidationState() {
+		return reconcile.Result{RequeueAfter: pollingPeriod}, nil
+	}
+
+	result, err := c.ProcessCluster(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("deprovisioning cluster, %w", err)
+	} else if result == ResultRetry {
+		return reconcile.Result{Requeue: true}, nil
+	} else if result == ResultNothingToDo {
+		c.lastConsolidationState = c.cluster.ClusterConsolidationState()
+	}
+	return reconcile.Result{RequeueAfter: pollingPeriod}, nil
+}
+
+// candidateNode is a node that we are considering for deprovisioning along with extra information to be used in
+// making that determination
+type candidateNode struct {
+	*v1.Node
+	instanceType   cloudprovider.InstanceType
+	capacityType   string
+	zone           string
+	provisioner    *v1alpha5.Provisioner
+	disruptionCost float64
+	pods           []*v1.Pod
+}
+
+// ProcessCluster is exposed for unit testing purposes
+// ProcessCluster loops through implemented deprovisioners
+// nolint:gocyclo
+func (c *Controller) ProcessCluster(ctx context.Context) (Result, error) {
```

Comment:
I agree with you since this makes sense to me if there's only one deprovisioner, which since the emptiness and consolidation settings are mutually exclusive, it basically is. The only problem is with adding in expiration and implementing other deprovisioners, we need to be able to distinguish the returned results of `NothingToDo` and `Success`, which both would return `(reconcile.Result{}, nil)` right now. WDYT about keeping as is, since Expiration will likely need this distinguish this?